Hierarchy of Sized traits (#3729)
rust-lang/rfcs: text/3729-sized-hierarchy.md

- Feature Name: `sized_hierarchy`
- Start Date: 2024-09-30
- RFC PR: [rust-lang/rfcs#3729](https://github.com/rust-lang/rfcs/pull/3729)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

All of Rust's types are either *sized*, which implement the `Sized` trait and
have a statically known size during compilation, or *unsized*, which do not
implement the `Sized` trait and are assumed to have a size which can be computed
at runtime. However, this dichotomy misses two categories of type - types whose
size is unknown during compilation but is a runtime constant, and types
whose size can never be known. Supporting the former is a prerequisite to stable
scalable vector types and supporting the latter is a prerequisite to unblocking
extern types.

This RFC necessarily depends on the experimental feature
[`const Trait`][goal_const_traits] and is written assuming familiarity with that

> feature.

Reviewer: I thought I knew the feature quite well, but it is not clear to me why traits need to be specially marked to allow "const" trait bounds. Is that explained anywhere? Sadly the link placed here doesn't go anywhere that explains anything.

Author: I believe it's just that if a trait isn't marked as const then it doesn't permit const implementations and as such it doesn't make sense to have const bounds on it, but I'm likewise making educated guesses from what I've read and understood from speaking to those working on const traits.

registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.

As an alternative to releasing SIMD extensions with greater bit widths, Arm and
RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
Extension/RVV respectively) where the bit width of vector registers depends on the
CPU implementation, and the instructions which operate these registers are bit
width-agnostic.

As a consequence, these types are not `Sized` in the Rust sense, as the size of
a scalable vector cannot be known during compilation, but is a runtime constant.
For example, the size of these types could be determined by inspecting the value
in a register - this is not available at compilation time and the value may
differ between any given CPU implementation. Both SVE and RVV have mechanisms to
change the system's vector length (up to the maximum supported by the CPU
implementations) but this is not supported by the proposed ABI for these types.

However, despite not implementing `Sized`, these are value types which should
implement `Copy` and can be returned from functions, can be variables on the

> stack, etc. These types should implement `Copy` but given that `Copy` is a
> supertrait of `Sized`, they cannot be `Copy` without being `Sized`, and

Reviewer 2: ```suggestion
stack, etc. These types should implement `Copy` but given that `Sized` is a
supertrait of `Copy`, they cannot be `Copy` without being `Sized`, and
```

It's Copy \< Clone \< Sized, not Sized \< Copy \< Clone.

they aren't `Sized`.

> Introducing a `const Sized` trait will enable `Copy` to be implemented for

Reviewer 3: The second one sounds better and more precise to me.

> Introducing a `const Sized` trait will enable `Copy` to be implemented for

Author: Fixed!

those types whose size is a runtime constant to function correctly without
special cases in the type system. See
[rfcs#3268: Scalable Vectors][rfc_scalable_vectors].

## Unsized types and extern types
[unsized-types-and-extern-types]: #unsized-types-and-extern-types

[Extern types][rfc_extern_types] has long been blocked on these types being
neither `Sized` nor `?Sized` ([relevant issue][issue_extern_types_align_size]).

RFC #1861 defined that `std::mem::size_of_val` and `std::mem::align_of_val`
should not be defined for extern types but not how this should be achieved, and
suggested an initial implementation could panic when called with extern types,
but this is always wrong, and not desirable behavior in keeping with Rust's
values. `size_of_val` returns 0 and `align_of_val` returns 1 for extern types in

> the current implementation. Ideally `size_of_val` and `align_of_val` would error

Reviewer: That's not correct, I made them panic a while ago: https://github.com/rust-lang/rust/pull/118534

Author: Fixed!

if called with an extern type, but this cannot be expressed in the bounds of
`size_of_val` and `align_of_val` and this remains a blocker for extern types.

Furthermore, unsized types cannot be members of structs as their alignment is
unknown and this is necessary to calculate field offsets. Extern types also
cannot be used in `Box` as `Box` requires size and alignment for both allocation

> and deallocation.

Reviewer: We currently allow extern types to appear as the first field of a struct. There's some unstable code relying on that, even in the compiler itself.

Author: Fixed!

Author: I think you meant last field though, that's what it seemed like was true from some quick testing, and that's what I've written. Happy to change it if I'm mistaken.

Reviewer: It also has to be the last field as usual for unsized types. But that's not enough, since as you said in general, computing the offset requires knowing the alignment.

The current entirely ad-hoc solution to this is to panic during the offset computation of the offset is not `0`. If it is `0`, then the alignment does not matter for offset computation.

[guide-level-explanation]: #guide-level-explanation
Most types in Rust have a size known at compilation time, such as `u32` or
`String`. However, some types in Rust do not have known sizes.

For example, slices have an unknown length while compiling and are
known as *dynamically-sized types*, their size must be computed at runtime.
There are also types with a runtime constant size (but unknown at compile time),
and *unsized types* with no size whatsoever.

Various parts of Rust depend on knowledge of the size of a type to work, for
example:

- [`std::mem::size_of_val`][api_size_of_val] computes the size of a value,
  and thus cannot accept extern types which have no size, and this should
  be prevented by the type system.
- Rust allows dynamically-sized types to be used as the final field in a struct,
  but the alignment of the type must be known, which is not the case for extern
  types.
- Allocation and deallocation of an object with `Box` requires knowledge of
  its size and alignment, which extern types do not have.
- For a value type to be allocated on the stack, it needs to have constant
  known size, which dynamically-sized and unsized types do not have (but 

>   sized and "runtime-sized" types do).

Reviewer 2: We might want to clarify that technically speaking dynamic stack allocation *exists* (cc `alloca`, C's VLAs) just not in Rust (barring incomplete features like `unsized_locals`, `unsized_fn_params`).

Author: Added some wording about this.

it may additionally require a value  (therefore `size_of` is not implemented for
`ValueSized`, only `size_of_val`). For example, `[usize]` implements
`const ValueSized` as knowing the type and target is not sufficient, the number of
elements in the slice must also be known, which requires having the value.

As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
requires a value, knowledge of the type and target platform, and can only be
computed at runtime. For example, `[svint8_t]` requires a value to know how
many elements there are, and then information from the runtime environment
to know the size of a `svint8_t`.

`Pointee` is implemented by any type that can be used behind a pointer, which is
to say, every type (put otherwise, these types may or may not be sized at all).
For example, `Pointee` is therefore implemented on a `u32` which is trivially
sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
no known size.

All type parameters have an implicit bound of `const Sized` which will be

> automatically removed if a `Sized`, `const ValueSized`, `ValueSized` or

Reviewer 2: presumably plus `~const Sized`, `~const ValueSized`?

What happens on `?ValueSized` and `?Pointee`? Remember that `?Trait` is currently permitted for all traits `Trait` (and leads to a non-lint warning if `Trait` ≠ `Sized`)

<sub>if this paragraph were to get updated, the corresponding one in section *`Sized` bounds* should be updated accordingly</sub>

Author: Fixed! I've noted that `?ValueSized` and `?Pointee` would be ignored and warn like other traits.

    │ {type, target} │     implies      │ {type, target, runtime env} │
    └────────────────┘                  └─────────────────────────────┘
            │                                          │
         implies                                    implies
            │                                          │
            ↓                                          ↓
┌───────────────────────┐             ┌────────────────────────────────────┐
│ const ValueSized      │ ──────────→ │ ValueSized                         │
│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
└───────────────────────┘             └────────────────────────────────────┘
                                                       │
                                                    implies
                                                       │
                               ┌───────────────────────┘
                               ↓
                     ┌──────────────────┐
                     │ Pointee          │
                     │ {runtime env, *} │
                     └──────────────────┘
```

Or, in Rust syntax:

> ```rust=

Reviewer 2: ?

````suggestion
```rust
````

and similar instances throughout this document

Author: This is just a HackMD-ism, it adds line numbers to the HackMD rendering, I'll remove it.

Author: Fixed these now.

> const trait Sized: ~const ValueSized {}

Reviewer 2: Side note: `const trait Trait {}` syntax is not a thing (yet), it's `#[const_trait] trait Trait {}` at the moment under `const_trait_impl`. Of course, this can stay as is if it's meant to be *conceptual*.

Author: Updated this to use the current syntax.

> const trait Sized: ~const ValueSized {}

Author: As per https://github.com/rust-lang/rfcs/pull/3729#discussion_r1845082704, I don't think it makes sense for the RFC to say anything about this.

> const trait ValueSized: std::ptr::Pointee {}

Reviewer 4: Note that this is a breaking change: It makes it ambiguous whether `T::Metadata` refers to the assoc type of `Pointee` or an user-provided trait:

```rust
trait Foo {
    type Metadata;
}

fn foo<T>() -> T::Metadata // does this refer to Foo::Metadata or Pointee::Metadata?
where
    T: Foo /* + Sized + ValueSized + Pointee*/
{
    todo!()
}
```

This can break [real code](https://github.com/search?q=lang%3Arust+%2F%28%3F-i%29%5BA-Z%5D%3A%3AMetadata%2F+&type=code) and should be addressed by this RFC.

Reviewer 5: This is also why `Sized` (as it exists today) cannot have a `const SIZE: usize` associated const. Perhaps there is need for a `#[bikeshed_hide_associated_item]` attr applied to trait assocaited items so they can only be accessed using fully qualified syntax `<T as Pointee>::Metadata`; or perhaps the attr could be rustc-internal and indicate an edition before which the item is hidden as such.

Author: This is a good catch, I hadn't considered this. A simple alternative that earlier versions of the proposal had was just to introduce a new `Pointee` marker trait rather than re-use `std::ptr::Pointee`.

Reviewer: There was anyway a proposal somewhere to expose the metadata type as `Metadata<Pointee>` rather than `<T as Pointee>::Metadata`, this would be another argument in favor of that.

Author: I've updated the RFC to make `Pointee` a new marker trait.

supertraits of `Sized` and `Sized` being a default bound, these
backwards-incompatibilities are avoided for `ValueSized` and `Pointee`.

Relaxing a bound from `Sized` to `ValueSized` or `Pointee` is non-breaking as
the calling bound must have either `T: Sized` or `T: ?Sized`, both of which
would satisfy any relaxed bound[^3].

However, it would still be backwards-incompatible to relax the `Sized` bound on
a trait's associated type[^4] for the proposed traits.

It is possible further extend this hierarchy in future by adding new traits between
those proposed in this RFC or after the traits proposed in this RFC without breaking
backwards compatibility, depending on the bounds that would be introduced[^5].

[^1]: Adding a new automatically implemented trait and adding it as a bound to
      an existing function is backwards-incompatible with generic functions. Even
      though all types could implement the trait, existing generic functions will be
      missing the bound.

      If `Foo` were introduced to the standard library and implemented on every
      type, and it was added as a bound to `size_of` (or any other generic
      parameter)..

      ```rust=

>       auto trait Foo;

Reviewer 2: ```suggestion
      auto trait Foo {}
```

and second occurrence below

Author: Fixed!

      - After `Pointee`
        - i.e. `Sized: ValueSized: Pointee: NewSized`
        - For the same reasons as the traits proposed in this RFC, adding bounds
          on a new trait after `Pointee` is backwards compatible (except for on a
          trait's associated type). Any existing function will have stricter
          bounds than the new trait and so could be relaxed to the new trait
          without breaking existing code.

## `Sized` bounds
[sized-bounds]: #sized-bounds

`?Sized` would be made syntactic sugar for a `const ValueSized` bound. A
`const ValueSized` bound is equivalent to a `?Sized` bound as all values in Rust
today whose types do not implement `Sized` are valid arguments to
[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
computed given a value and knowledge of the type and target platform, and
therefore will implement `const ValueSized`. As there are currently no
extern types or other types which would not implement `const ValueSized`,
every type in Rust today which would satisfy a `?Sized` bound would satisfy
a `const ValueSized` bound.

**Edition change:** In the current edition,`?Sized` will be syntatic sugar for

> a `const ValueSized` bound. In the next edition, use of `?Sized` syntax will be prohibited

Reviewer 2: > `?Sized` syntax will be prohibited

What about `?Trait` (`Trait` ≠ `Sized`). Feels a bit weird to reject `?Sized` in Rust N and continue to permit all other (no-op) `?Trait` bounds.

but only at runtime, which requires ensuring these functions are only const if their
arguments have const implementations of the relevant sizedness traits.

[`size_of`][api_size_of] is a const-stable function since Rust 1.24.0 and currently
accepts a `T: Sized` bound. Therefore, even when used in a const context, `size_of`
could accept a runtime-sized type. It is therefore necessary to modify the bounds of
`size_of` to accept a `T: ~const Sized`, so that `size_of` is a const function
if-and-only-if `Sized` has a `const` implementation.

```rust=
pub const fn size_of<T: ~const Sized>() -> usize {
    /* .. */
}
```

This has the potential to break existing code like `uses_size_of` in the below
example. However, due to the changes described in [`Sized` bounds][sized-bounds]
(changing the implicit `T: Sized` to `T: const Sized`, and treating explicit
`T: Sized` bounds as `T: const Sized` in the current edition), this code would
not break:

```rust=
fn uses_size_of<T: Sized>() -> usize {
    const { std::mem::size_of<T>() }
}

fn another_use_of_size_of<T: Sized>() -> [u8; size_of::<T>()] {
    std::array::repeat(0)
}
```

> [`size_of_val`][api_size_of_val] is also const-stable, so like `size_of` above,

Reviewer: Not yet. FCP is still ongoing in https://github.com/rust-lang/rust/issues/46571. We can halt that if we have to.

Author: Yeah, I wrote this assuming that FCP goes ahead.

> its bound should be changed to `T: ~const Sized` and this would not result in any

Reviewer 6: ```suggestion
its bound should be changed to `T: ~const ValueSized` and this would not result in any
```

Author: Fixed!

   summarizes these discussions)
2. Keep `?Sized`'s existing meaning of removing the implicit `Sized` bound
      
   This is the only viable option, but this would complicate changing
   `size_of_val`'s existing `?Sized` bound:

   Without `Pointee`, `?Sized` would be equivalent to `const ValueSized` until
   extern types are stabilised (e.g. a `?Sized` bound would accept exactly the
   same types as a `const ValueSized` bound, but after extern types are introduced,
   `?Sized` bounds would accept extern types and `const ValueSized` bounds would not).
   extern types would need to be introduced over an edition and all existing `?Sized`
   bounds rewritten to `?Sized + const ValueSized`. This is the same mechanism described
   in [rfcs#3396][rfc_extern_types_v2] to introduce its `MetaSized` trait.

## Why use const traits?
[why-use-const-traits]: #why-use-const-traits

Previous iterations of this RFC had both linear[^6] and non-linear[^7] trait hierarchies
which included a `RuntimeSized` trait and did not use const traits. However, both of
these were found to be backwards-incompatible due to being unable to relax the

> supertrait of `Clone`. Without const traits, it is not possible to represent

Reviewer 7: This is mentioned in the top-level RFC description:

> This RFC relies on experimental, yet-to-be-RFC'd const traits, so this is blocked on that. I haven't squashed any of the previous revisions but can do so if/when this is approved.

`const` traits have a *huge* design and implementation space, and it seems like stabilization of such a feature is quite a ways off.

Have you considered whether there's a way, possibly via trait aliases or similar, to create a stable interface for the new `Sized` hierarchy without first stabilizing the `const` trait feature?

Reviewer 8: A sufficiently minimal subset of the `const` traits implementation space that fully enables this proposal may actually be possible to drive to stabilization in 2025, such that I do not think this RFC should be hung on this concern.

I defer to @fee1-dead, @oli-obk, and @compiler-errors on this, of course.

Reviewer 8: Also, trait aliases aren't stable either.

Author: From speaking to @compiler-errors, I'm under the impression that the types team have made a lot of progress with the implementation of const traits and feel quite confident in the implementation and that they understand the feature and its implications. I'm sure there's still lots to do but it sounded promising.

I don't think it's possible to implement this without const traits - we considered introducing another trait in the hierarchy `RuntimeSized` in earlier revisions, but it isn't possible to relax the supertrait of `Clone` from `Sized` to `RuntimeSized`, so you need to keep it as `Sized` and rely on something like const traits.

Reviewer 7: To clarify, I'm wondering if it's possible to rely on const traits in a way that doesn't expose to end-users that they're using const traits. This would potentially allow stabilization of the feature before broad stabilization of a (much larger) const traits feature.

Reviewer 6: why not have `const Sized` under a separate feature gate from general const traits and stabilize just `const Sized` first?

Reviewer 2: because we probably also want `~const Sized` which is officially placeholder syntax

Reviewer 6: why not just say `~const Sized` can wait till later, since `const Sized` is needed to express what `Sized` does now, but `~const Sized` is new?

Reviewer 9: The syntax part may be separatable, but the background work (ensuring everything is consistent and forward compatible) required is the same

Author: Even if you had only `const Sized`, you would probably want to be able to write functions with `~const Sized` bounds so that when invoked out of a const context, they can accept runtime-sized types.

which use extern types as values (e.g. such as taking an extern type as an argument),
and instead rely on it being impossible to get a extern type that is not behind a
pointer or a reference. This also implies that `ValueSized` types can be used as values,
which would remain prohibited behind the `unsized_locals` and `unsized_fn_params`
features until these are stabilised.

With these changes to the RFC, it would be possible to support wasm's `externref` and
opaque types from some GPU targets.

## Alignment
[alignment]: #alignment

There has been community interest in an [`Aligned` trait][rfc_aligned] and there
are examples of `Aligned` traits being added in the ecosystem:

- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
  whether a type has an alignment or not.

Furthermore, the existing incorrect behaviour of `align_of_val` returning one is
used in `rustc` with the [`OpaqueListContents` type][rustc_opaquelistcontents] to
implement a custom DST. This use case would break with this RFC as written, as

> the alignment of an extern type would correctly be undefined.

Reviewer: This is not correct, since align_of_val does not return 1 any more for extern types.

Author: Fixed!

└───────────────────────┘             └────────────────────────────────────┘
                                                       │
                                                    implies
                                                       │
                               ┌───────────────────────┘
                               ↓
                     ┌──────────────────┐
                     │ Pointee          │
                     │ {runtime env, *} │
                     └──────────────────┘
                               │
                            implies
                               │
                               ↓
                     ┌──────────────────┐
                     │ Value            │
                     │ {runtime env, *} │
                     └──────────────────┘
```

`Pointee` is still defined as being implemented for any type that can be used
behind a pointer and may not be sized at all, this would be implemented for
effectively every type except wasm's `externref` (or similar opaque types from
some GPU targets). `Value` is defined as being implemented for any type that can

> be used as a value, which is all types, and also may not be sized at all.

Reviewer 10: One possible complication: Type which are `Value` (but not `Pointee`, referred to as `!Pointee` types from now on) would be _**very**_ restrictive in how they can be used (without additional, IMO somewhat significant, changes to the language).

If a type is neither `Copy` nor allows references to it to be taken, then it can be basically used in a single function and passed into a single function, but not much more (unless it's returned again by the called function, which might work as long as you only need to return a single value).

To make any real use of`!Pointee` types, I think the language would need to support at least one (or more) of the following:

1) Allow returning multiple `!Pointee` values from a function, potentially by supporting `!Pointee` types in structs / tuples (which would require passing each field of such aggregates individually at the ABI level).

2) Allow such `!Pointee` types to be copied, either by somehow allowing them to implement `Copy` (despite the conflicting super traits), or allowing them to be copied even if they do not implement the `Copy` trait.

3) Allow **immutable** references to `!Pointee` values to exist, but make these themselves `!Pointee`. At the ABI level such references would be represented by the value itself. `UnsafeCell` would not be allowed to contain `!Pointee`.

(3) would be the most flexible solution here, though also the largest addition to the language semantics, I think.

Anyway, the details here are out of scope of this RFC, I think, but to me this seems like a significant enough complication that it might be worthwhile to mention here, or to weaken the wording ("proposed hierarchy of traits **could support** this", "it would be **possible to support**") a bit, to not imply that this would solve all the `externref` problems. (Maybe adding something along the lines of "though making proper use of `!Pointee` types could require additional additions to the language" somewhere might be sufficient for this RFC).

Reviewer 10: I've been thinking more about this (see also https://hackmd.io/@TimNN/Hk1LO6iMJe) and I think one thing that doesn't work well at all is the modeling of the "can be stored in a local variable" property.

We currently have four groups of types that could potentially be stored in local variables:

* Sized values
* Dynamically-sized values
* Unsized values (i.e. `extern` types)
* Abstract values, as I have taken to calling them (e.g. `externref`).

`Sized` values are fully supported.

Dynamically-sized values are support via `unsized_locals` and `unsized_fn_params`. They use a variable amount of storage at runtime. They (currently) cannot be used as a function return value.

Unsized values conceptually cannot be stored in local variables. (This RFC currently proposes that  _"it could instead be permitted to write functions which use extern types as values, and instead rely on it being impossible to get a extern type that is not behind a pointer or a reference"_, but that doesn't seem like a great solution to me).

Abstract values should be usable in local variables, function arguments, and as function return values. Contrary to dynamically-sized values, they do not use a variable amount of storage at runtime (but have other restrictions).

---

I don't have a good solution right now, especially since I'm not sure about what the long-term plan for dynamically-sized values is (should they be usable as function return values?).

However I'm currently feeling that the "can be stored in a local variable" property might not fit very well into the hierarchy proposed here.

If we assume that dynamically-sized values support all operations that are supported for abstract values (including being usable as a return value), then it might be reasonable to define `ValueSized: Pointee + AbstractValue`, with no direct relationship between `Pointee` and `AbstractValue`

(Though if `Pointee` and `AbstractValue` exist independently, then the question becomes "what does `: ?ValueSized` mean?". Depending on the outcome of other discussions, this may mean bounds such as`: ?ValueSized + AbstractValue`).

--- 

One way I've been trying to think about this is, what would be appropriate bounds for `V, R and Self` in the following traits?

```rs
trait FromVal<V> { fn from_val(_: V) -> Self }

trait FromRef<R> { fn from_ref(_: &R) -> Self }
```

registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.

As an alternative to releasing SIMD extensions with greater bit widths, Arm and
RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
Extension/RVV respectively) where the bit width of vector registers depends on the
CPU implementation, and the instructions which operate these registers are bit
width-agnostic.

As a consequence, these types are not `Sized` in the Rust sense, as the size of
a scalable vector cannot be known during compilation, but is a runtime constant.
For example, the size of these types could be determined by inspecting the value
in a register - this is not available at compilation time and the value may
differ between any given CPU implementation. Both SVE and RVV have mechanisms to
change the system's vector length (up to the maximum supported by the CPU
implementations) but this is not supported by the proposed ABI for these types.

However, despite not implementing `Sized`, these are value types which should
implement `Copy` and can be returned from functions, can be variables on the

> stack, etc. These types should implement `Copy` but given that `Sized` is a

Reviewer 11: How does stack allocation works for the "runtime constant" types, at binary level? Is it different from dynamic stack allocation?
Does it work for SVE specifically, or it's possible to do for other "runtime constant" types too?

I assume that the "runtime constant" value becomes available somewhere before the call to `main`?
Or earlier (e.g. during linking), and you need to specify the "runtime env" during the build?

Reviewer 12: > How does stack allocation work for the "runtime constant" types, at binary level? Is it different from dynamic stack allocation?
> Does it work for SVE specifically, or it's possible to do for other "runtime constant" types too?

~~I believe it's just "dynamic `alloca`", I don't recall ever seeing anything in LLVM about "runtime constant" stack frame sizes.~~
(while searching for examples to add elsewhere in this comment I've found that while LLVM `alloca` *is* used, [it's more "typed" than I expected](https://github.com/search?q=repo%3Allvm%2Fllvm-project+%2Fstack-id%3A+scalable-vector%7CTargetStackID%3A%3AScalableVector%2F&type=code), and the stack frame layout is aware of the distinction - cc @nikic, not sure how this will look once `alloca Type` is gone)

<sub>While it would be neat to treat such values (and those derived from them with simple arithmetic) as "relocation-time constants" and have the dynamic linker patch them into e.g. as many instruction immediates as possible, [I have recently argued against the desirability of storing vector registers to memory](https://github.com/rust-lang/rfcs/pull/3268#issuecomment-2468886230) (outside of e.g. context switching and other low-level tasks), so I'm skeptical of the value of relocation-level support.</sub>

---

> I assume that the "runtime constant" value becomes available somewhere before the call to `main`?

In theory, some mechanism like ELF auxv could be used, but AIUI it's just an instruction reading a "special register":
- ARM has `mrs`/`msr` instructions to move between general-purpose and special registers
  - SVE also has [`rdvl`](https://developer.arm.com/documentation/ddi0602/2024-09/SVE-Instructions/RDVL--Read-multiple-of-vector-register-size-to-scalar-register-) and similarly helpful instructions
  - combined with `alloca`, [Clang/LLVM-produced asm gets wild](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-alloca.ll) (Ctrl+F `vl`)
- RISC-V more specifically calls them "CSR"s ("Control and Status Register"s)
  - the relevant `size_of` value for RVV, `vlenb`, is a read-only CSR
  - (while RVV has other CSRs and instructions, see below why they're not relevant for `size_of`)

I'd compare it to using x86 `cpuid` to e.g. determine the SIMD extension with the largest registers (and it similarly is "not supposed to change during execution" AFAICT).

LLVM seems to abstract this concept in its [`@llvm.vscale.iN`](https://llvm.org/docs/LangRef.html#llvm-vscale-intrinsic) intrinsic, documented as:
> `vscale` is a positive value that is constant throughout program execution, but is unknown at compile time.
  - anything more fine-grained than `vscale` has to go either through target-specific intrinsics *or* the newer [predicated vector intrinsics](https://llvm.org/docs/LangRef.html#int-vp), which take *both* a mask *and* an "EVL" ("Effective Vector Length") and in theory could (eventually) be useful for a `core::simd`-like abstraction

---

There's also Linux-specific syscalls/documentation:
- SVE: https://docs.kernel.org/arch/arm64/sve.html
  - `prctl(PR_SVE_GET_VL)` as alternative to getting VL through instructions
    - [not widely used AFAICT](https://github.com/search?q=prctl%28PR_SVE_GET_VL%29+-path%3A%2Fstress%7Cvalidation%7Ctest%7C%28%5C.%28rst%7Cmd%7Ctxt%7Chtml%29%24%29%2F&type=code) (outside the Java HotSpot VM, for JIT reasons I assume) and [Clang/LLVM definitely use `rdvl`](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-vscale.ll#L8-L15) and [related instructions](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-alloca.ll)
  - can also set current VL (or defer to `execve`, i.e. set it for a child process)
  - not expecting changing VL on the fly to be compatible with code using SVE C intrinsics
- RVV: https://docs.kernel.org/arch/riscv/vector.html
  - less control than SVE, as RVV only has these two lengths:
    - `vlenb`: the *hardwired* µarch implementation maximum (which is what non-`const` `size_of` would supposedly return, [much to my surprise](https://github.com/rust-lang/rfcs/pull/3268#issuecomment-2468886230))
    - `vl`: the dynamic length, meant to be set by intrinsics and auto-vectorized loops *on the fly* (to remove the need for any scalar loops, as e.g. the last iteration of the vector loop can still work with `vl` as low as `1`)
  - > To get the availability of V in an ELF program, please read `COMPAT_HWCAP_ISA_V` bit of `ELF_HWCAP` in the auxiliary vector.
    - so there is something available through ELF auxv, but it's just extension bits (and ARM has them too)

Out of curiosity, I've also looked at the ELF auxv HWCAP for x86 and AFAICT it's [`cpuid(eax=1).edx`](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/arch/x86/include/asm/elf.h#L241-L254) which [sadly only goes up to SSE2 in terms of extension bits](https://en.wikipedia.org/wiki/CPUID#EAX.3D1:_Processor_Info_and_Feature_Bits) (newer ones being in ECX), and [HWCAP2 is a couple of kernel-controlled features](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/arch/x86/include/uapi/asm/hwcap2.h).

---

Also, while looking up the LLVM-specific examples to link above, I came across SME/SVE2 "streaming mode" [which is a whole new can of worms](https://llvm.org/docs/AArch64SME.html), and I hope doesn't require even stranger types.

Reviewer 13: > (while searching for examples to add elsewhere in this comment I've found that while LLVM `alloca` _is_ used, [it's more "typed" than I expected](https://github.com/search?q=repo%3Allvm%2Fllvm-project+%2Fstack-id%3A+scalable-vector%7CTargetStackID%3A%3AScalableVector%2F&type=code), and the stack frame layout is aware of the distinction - cc @nikic, not sure how this will look once `alloca Type` is gone)

I'd expect that `alloca` would accept a [`TypeSize`](https://llvm.org/doxygen/classllvm_1_1TypeSize.html), which is either a fixed size or a size multiplied by vscale. I think that would cover it?

Reviewer 14: This thread has made me much more sceptical that these types provide good motivation for this feature. It's interesting that `externref` style types have similar requirements: they need to be `Copy`, they can be passed as arguments, returned as results, used as local variables, etc, but importantly they aren't `Sized` (or even `ValueSized`)... Similarly, references to both types are weird because they aren't pointers (or at least shouldn't be in the case of vector types as I think @eddyb is suggesting).

I haven't yet come up with a productive solution though...

Author: > Similarly, references to both types are weird because they aren't pointers (or at least shouldn't be in the case of vector types as I think @eddyb is suggesting).

As I understand it, scalable vectors can be used behind pointers, only `externref` style types cannot.

As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
requires a value, knowledge of the type and target platform, and can only be
computed at runtime. For example, `[svint8_t]` requires a value to know how
many elements there are, and then information from the runtime environment
to know the size of a `svint8_t`.

`Pointee` is implemented by any type that can be used behind a pointer, which is
to say, every type (put otherwise, these types may or may not be sized at all).
For example, `Pointee` is therefore implemented on a `u32` which is trivially
sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
no known size.

All type parameters have an implicit bound of `const Sized` which will be
automatically removed if a `~const Sized`, `Sized`, `const ValueSized`,
`~const ValueSized`, `ValueSized` or `Pointee` bound is present instead.

Prior to the introduction of `ValueSized` and `Pointee`, `Sized`'s implicit bound
(now a `const Sized` implicit bound) could be removed using the `?Sized` syntax,

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Reviewer 6: this reminds me of C++ classes where you can figure out the size of the dynamic type by reading the vtable pointer, but, unlike Rust, that requires dereferencing the data part of the pointer to the type, whereas in Rust that info is passed as the pointer's metadata. So maybe we need both `ValueSized` and `RefSized` where for `RefSized` you can pass any old pointer to the type (since the metadata of a pointer must be valid but the data pointer doesn't need to), but for `ValueSized` you have to be able to dereference the data pointer, so takes something like `&T` but without the aliasing guarantees.

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Reviewer 3: > I don't think this is true, specifically `Mutex<T: ValueSized>` cannot be `ValueSized` as conceptually it would require locking the mutex in order to call `size_of_val` on the wrapped type. That feels terrifying and it's currently observable that's not the case because it doesn't deadlock when you call `size_of_val` while holding the Mutex.

Could you give the full code example of what you have in mind?  I want to check that what I have in mind here exactly matches what you do.

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Reviewer 14: This is a tad conceptual as it requires types that don't exist in rust as it exists today, but would exist with something like custom DSTs. Imagine this:
```rust
// Magic syntax that means that CStr isn't Sized
struct CStr(..);

impl ValueSized for CStr {
    // Not being proposed here but could exist
    // with custom DSTs.
    fn size_of_val(&self) -> usize {
        let data = self as *const u8;
        // Find first null byte and return length
    }
}
```
As you can see this type has to inspect the data behind the `&self` pointer in order to determine its size. (Admittedly I'm not entirely convinced we'd ever want this in rust because it feels like a performance foot gun, but this matches the semantics of `ValueSized` as described in the RFC)

Now consider `size_of_val(&Mutex<CStr>)` the only way that function can execute is if the `size_of_val` implementation for Mutex acquires a lock on the inner data, this is bad and it's observably not the case because this code doesn't deadlock: 
```rust
let mutex = Mutex::new(7);
let _guard = mutex.lock().unwrap();
size_of_val(&mutex);
```

The issue is that `ValueSized` as described doesn't match current rust's rules, where a type is only allowed to access the pointer metadata in order to determine its size. Changing the semantics to be that would mean that `CStr` couldn't implement `ValueSized`, meaning that `size_of_val(&Mutex<CStr>)` would throw a compile error as `Mutex<CStr>` also wouldn't implement `ValueSized`.

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Author: I don't think this interaction with mutexes and `size_of_val` would be that surprising.

If you have a type that you know requires a value to compute its size then computing the size of a mutex containing it would require the mutex lock itself. If we had reason to believe it was common for generic code that always works today to start deadlocking when instantiated with a `ValueSized` type, then that would give me pause, but otherwise I don't think this is too bad, it's just the natural interaction of mutexes and value-sized types. That said, I do think that `MetaSized` is a perfectly acceptable alternative that captures everything that we want from `ValueSized` (to the best of my knowledge at least).

I've added an alternative that describes how `MetaSized` could be included in this RFC instead of `ValueSized` if this were an issue we wanted to avoid.

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Reviewer 4: I don't think the mutex example highlights the importance of `MetaSized` enough if we ever want to allow custom code in `size_of_val` (such as locking a mutex).

On stable rust it is currently possible to call `size_of_val` on `&UnsafeCell<dyn Trait>`, so this must continue to compile.

If we were to define `T: ValueSized` as "the size of `T` can be determined from a value of type `&T`", then there is no way to express the (compiler-builtin) impl of `ValueSized` for `UnsafeCell` without `MetaSized`.

For most normal types this impl would be:
```rust
impl const ValueSized for Foo where tail(Foo): ~const ValueSized {}
```

However, for `UnsafeCell` this would be incorrect:
```rust
impl<T> const ValueSized for UnsafeCell<T> // the size of UnsafeCell<T> can be known from a value of type &UnsafeCell<T> ...
where T: ~const ValueSized // ... if the size of T can be known from a value of type &T
{}
```

The size of `UnsafeCell<T>` **cannot** be determined by a value of type `&UnsafeCell<T>` if computing the size from `&T` requires to run user code, because there is no way to safely obtain a reference to the inner type.

Instead, we need some other trait bound to correctly express under what circumstances we can obtain the size of an `UnsafeCell<T>` if we have a `&UnsafeCell<T>` and maintain backwards compatibility with stable rust -- this trait bound is exactly `T: MetaSized`! 
```rust
impl<T> const ValueSized for UnsafeCell<T> where T: ~const MetaSized {} // OK
```

> which is now equivalent to a `ValueSized` bound in non-`const fn`s and

Author: I've added more to this alternative's section based on your `UnsafeCell` example.

          bounds than the new trait and so could be relaxed to the new trait
          without breaking existing code.

## `Sized` bounds
[sized-bounds]: #sized-bounds

`?Sized` would be made syntactic sugar for a `const ValueSized` bound. A
`const ValueSized` bound is equivalent to a `?Sized` bound as all values in Rust
today whose types do not implement `Sized` are valid arguments to
[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
computed given a value and knowledge of the type and target platform, and
therefore will implement `const ValueSized`. As there are currently no
extern types or other types which would not implement `const ValueSized`,
every type in Rust today which would satisfy a `?Sized` bound would satisfy
a `const ValueSized` bound.

**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
a `const ValueSized` bound. As the `?Trait` syntax is currently accepted for any trait
but ignored for every trait except `Sized`, `?ValueSized` and `?Pointee` bounds would
be ignored. In the next edition, any uses of `?Sized` syntax will be rewritten to

> a `const ValueSized` bound. Any other uses of the `?Trait` syntax will be removed as

Author: I don't think this is possible, there is a future possibility in this RFC that very briefly touches on how we could relax associated types, but I don't think adding implicit `T:Assoc: const ValueSized` bounds would work, because something like this compiles today..

```
trait Foo {
    type Bar;
}

fn qux<T: Foo>() -> usize {
    std::mem::size_of::<<T as Foo>::Bar>()
}
```

..and wouldn't if we added that bound.

> a `const ValueSized` bound. Any other uses of the `?Trait` syntax will be removed as

Reviewer 3: The idea is that, if we relaxed the bound, it'd work like this:

```rust
trait Tr {
    type Ty: ?Sized default Sized;
    //       ~~~~~~~        ~~~~~
    //  The relaxed bound   |
    //                      ^ The backward compatible default for
    //                        use-site bounds.
}
```

Where at the use site then, if you wrote this...

```rust
fn f<T: Tr> {}
```

...it would be treated according to that `default` as though you had written:

```rust
fn f<T: Tr<Ty: Sized>> {}
//        ~~~~~~~~~~~
//      Added implicitly.
```

This works because loosening the bound doesn't break implementors.  The only reason it breaks callers is because it changes what they can assume by default without explicitly bounding the associated type.  But we could conceivably separate what callers can assume by default from the minimum that implementors are allowed to implement.

Then, conceivably, over an edition, we could change the default.  Use sites in older edition code would still get the old default.  During the edition migration, we would elaborate the code according to that default, which would work in both editions and preserve the semantics.

I asked around about this; unfortunately there are apparently some reasons it could be difficult to implement.

part of the migration and the `?Trait` syntax will be prohibited.

A default implicit bound of `const Sized` is added by the compiler to every type
parameter `T` that does not have an explicit `~const Sized`, `Sized`, `?Sized`,
`const ValueSized`, `~const ValueSized`, `ValueSized` or `Pointee` bound. It is
backwards compatible to change the current implicit `Sized` bound to an `const Sized`
bound as every type which exists currently will implement `const Sized`.

**Edition change:** In the current edition, all existing `Sized` bounds will be
syntatic sugar for `const Sized` (both explicitly written and implicit). In the next
edition, existing `Sized` bounds will be rewritten to `const Sized` bounds and new
bare `Sized` bounds will be non-const. `Sized` supertraits (such as that on `Clone`
or `Default`) will not be sugar for `const Sized`, these will remain bare `Sized`.
If traits with a `Sized` supertrait are later made const, then their supertrait
would be made `~const Sized`.

> An implicit `const ValueSized` bound is added to the `Self` type of traits. Like

Author: I've also added this as an example in the RFC.

> An implicit `const ValueSized` bound is added to the `Self` type of traits. Like

Reviewer 14: I disagree, the difference between having no default supertrait and having one is massive. Not having a default supertrait means that traits naturally apply to all types and library authors don't have to remember (or be asked) to relax that bound where possible (which would be a breaking change).

Not being able to implement any std traits on these types feels like a big problem to me, it would make them a nightmare to use without Debug, Clone, PartialEq, etc. Admittedly, this is probably less of a problem for the vector types as it's less clear (to me at least) what they could implement given they might contain uninitialised bytes. I wonder if it would be possible to relax the bounds on existing traits by introducing explicit bounds on traits with default implementations.

> An implicit `const ValueSized` bound is added to the `Self` type of traits. Like

Author: > I disagree, the difference between having no default supertrait and having one is massive.

I wasn't intending to argue that this limitation wasn't significant, just that I didn't think it was hard to teach or remember.

> An implicit `const ValueSized` bound is added to the `Self` type of traits. Like

Author: I've pushed another change related to this, I think the implicit bound on `Self` is necessary but it should be possible to relax it backwards-compatibily.

This is forward compatible with trait bounds which have sizedness supertraits
implying the removal of the default `const Sized` bound.

## Ecosystem churn
[ecosystem-churn]: #ecosystem-churn

It is not expected that this RFC's additions would result in much churn within
the ecosystem. Almost all of the necessary changes would happen automatically
during edition migration.

All bounds in the standard library should be re-evaluated during the
implementation of this RFC, but bounds in third-party crates need not be.

As runtime-sized types will primarily be used for localised performance optimisation,
and `Pointee` types will primarily be used for localised FFI, neither is expected
to be so pervasive throughout Rust codebases to the extent that all existing
`const Sized`, `~const ValueSized` or `ValueSized` bounds (after edition migration) would
need to be immediately reconsidered in light of their addition, even if in many cases
these could be relaxed.

If a user of a runtime-sized type or a `Pointee` type did encounter a bound that
needed to be relaxed, this could be changed in a patch to the relevant crate without

> breaking backwards compatibility as-and-when such cases are encountered.

Reviewer 14: I think this is an overly rosy outlook. It may be very annoying for types that aren't `const ValueSized` to be locked out of implementing traits from other crates as it will make it much harder for them to be used like normal types, also we'd need to teach crate authors to write new code as permissively as possible. 
Additionally it's not always going to be possible to relax these bounds without a breaking change, my personally scariest trait is [serde's `Serializer` trait](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_some) as it's impossible to relax that bound as serializers might rely on it but it prevents these new types from being first class members of the swede ecosystem.

Reviewer 3: Could you elaborate on this point a bit and spell it out further?  I feel like there may be something important here to consider, and am interested in seeing it detailed.

Reviewer 15: Here's a more self contained example consider a library with the following trait:

```rust

trait A {
  fn a<T: ?Sized>(&mut self, x: &T);
}

```

An impl of that trait in a different crate may call `size_of_val`,  so the crate that defines A can't backwards compatibly relax the type of T to Pointee instead of `const ValueSized`.

Reviewer 15: A problematic case of this in std is Deref. 

Ideally, I think that the Target type should be relaxed to Pointee, but that wouldn't be backwards compatible, because existing code might depend on being able to call `size_of_val` on `&<T as Deref>::Target`.

Reviewer 14: As mentioned [here](https://github.com/rust-lang/rfcs/pull/3729#discussion_r1851828543), I think that it is possible to relax associated types in a backwards compatible way, as long as we do it with the edition migration required for this RFC. However, I agree with your minimal example on a situation where I don't think there's any backwards compatible solution.

Author: This is a good catch, I didn't realise trait methods would an issue. I don't have a good answer to how to deal with those, but I've noted this in the RFC.

- [rfcs#2310: DynSized without ?DynSized][rfc_dynsized_without_dynsized] was
  proposed at a similar time and was similarly focused only on making `Sized` more
  flexible, but had a bunch of machinery for avoiding backwards incompatibility
  that this RFC believes is unnecessary. Like this proposal, it avoided making
  `DynSized` a default bound and avoided having a relaxed form of it.
    - However, this proposal didn't suggest removing default `Sized` bounds in
      the presence of other size trait bounds.
- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
  major inspiration for this proposal. It has everything this proposal has except
  for the const size traits and all the additional context an RFC needs.

Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
of the type system which can result in new implicit bounds or implicit supertraits
being infeasible for implementation reasons, but [it is believed that this is no
longer relevant][zulip_issue_regions_too_simplistic].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

None currently.

# Future possibilities

> [future-possibilities]: #future-possibilities

Reviewer 16: I am not smart/informed enough to tell for sure, but I _feel_ like if we want to change the whole default `?Sized` thing, and necessarily deal with the breakage/fallout from that, we might want to consider adding `Move` at the same time?

Anyhow, just a consideration, excellent RFC, full support from my side!

Reviewer 17: Be still my heart. 

I guess this could just be the Baader-Meinhof phenomenon at work, but I feel like `Move` is coming up a lot recently. Do you know if anybody has put together recent summary of the "state of `Move`" — i.e. plausibility of ever having it even theoretically, would it require an edition, scale of the challenge, etc.?

Author: I haven't read enough of the context of `Move` to know whether it makes sense to incorporate into this proposals, but I assume it doesn't. If someone wants to go and summarise the history of those proposals and make a compelling case for how it could be integrated with this work then I'd be open to add it as a alternative that the language team could consider.

Reviewer: This RFC is changing the `Sized` axis of default bounds, an axis where we already have an existing default bound. I don't see how that would make it any easier to add an entirely new axis of default bounds.

be used as a value, which is all types, and also may not be sized at all.

Earlier in this RFC, `extern type`s have previously been described as not being
able to be used as a value, but it could instead be permitted to write functions
which use extern types as values (e.g. such as taking an extern type as an argument),
and instead rely on it being impossible to get a extern type that is not behind a
pointer or a reference. This also implies that `ValueSized` types can be used as values,
which would remain prohibited behind the `unsized_locals` and `unsized_fn_params`
features until these are stabilised.

With these changes to the RFC and possibility additional changes to the language, it
could be possible to support wasm's `externref` and opaque types from some GPU targets.

## Alignment
[alignment]: #alignment

There has been community interest in an [`Aligned` trait][rfc_aligned] and there
are examples of `Aligned` traits being added in the ecosystem:

- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
  whether a type has an alignment or not.

> An `Aligned` trait could be added to this proposal between `ValueSized` and `Pointee`

Reviewer 14: I think this is the wrong place to put it, `dyn Trait` is `ValueSized` but is not `Aligned`, but I can imagine extern types that are `Aligned` but not `ValueSized`. I think that means `Aligned` could be a super trait of `Sized` but doesn't otherwise fit in this hierarchy. (In fact I think there's an entirely separate hierarchy including `ValueAligned`, etc.)

Author: I've rewritten this section of the RFC.

can be made while preserving backwards compatibility, the following changes
could be made to the standard library:

- [`std::boxed::Box`][api_box]
    - `T: ?Sized` becomes `T: ValueSized`
    - It is not a breaking change to relax this bound and it prevents types
      only implementing `Pointee` from being used with `Box`, as these types
      do not have the necessary size and alignment for allocation/deallocation.
- [`std::marker::PhantomData`][api_phantomdata]
    - `T: ?Sized` becomes `T: Pointee`
    - It is not a breaking change to relax this bound and there's no reason why
      any type should not be able to be used with `PhantomData`.

As part of the implementation of this RFC, each `Sized`/`?Sized` bound in
the standard library would need to be reviewed and updated as appropriate.

# Drawbacks
[drawbacks]: #drawbacks

- This is a fairly significant change to the `Sized` trait, which has been in
  the language since 1.0 and is now well-understood.
- This RFC's proposal that adding a bound of `const Sized`, `const ValueSized`,
  `ValueSized` or `Pointee` would remove the default `Sized` bound is a significant

>   change from the current `?Sized` mechanism.

Reviewer: ```suggestion
  change from the current `?Sized` mechanism, and can be considered quite confusing.
```

Author: Fixed

compatible with this proposal as follows:

In the current edition, `?Sized` would need to be equivalent to `?Sized + const ValueSized` to
maintain backwards compatibility (see [the `size_of` and `size_of_val`
section][size_of-and-size_of_val] for rationale). In the next edition, `?Sized` would be
rewritten to `?Sized + const ValueSized` (remove the `Sized` default and add a `const ValueSized`
bound) and bare `?Sized` would only remove the `Sized` default bound.

Prior to the edition migration, the default bound is `Sized`, which could be changed using
the following syntax:

| With positive bounds | Keeping `?Sized` |
| -------------------- | ---------------- |
| `const Sized`        | `const Sized`    |
| `Sized`              | `Sized`          |
| `const ValueSized`   | `?Sized`         |
| `ValueSized`         | Not possible     |
| `Pointee`            | Not possible     |

After the edition migration, the default bound is `const Sized`, which could be changed
using the following syntax:

| With positive bounds  | Keeping `?Sized`                  |
| --------------------- | --------------------------------- |
| `const Sized`         | `const Sized`                     |

> | `Sized`               | `?const Sized + Sized`            |
> | `const ValueSized`    | `?const Sized + const ValueSized` |
> | `ValueSized`          | `?const Sized + ValueSized`       |
> | `Pointee`             | `?const Sized`                    |

Reviewer 3: ```suggestion
| `Sized`               | `?(const Sized) + Sized`            |
| `const ValueSized`    | `?(const Sized) + const ValueSized` |
| `ValueSized`          | `?(const Sized) + ValueSized`       |
| `Pointee`             | `?(const Sized)`                    |
```

Given the point that @RalfJung made (https://github.com/rust-lang/rfcs/pull/3729#issuecomment-2509011564), and given the ongoing parallel discussions about possible meanings for `?const`, it'd be better to disambiguate this.

Author: Fixed

> 

Reviewer: ```suggestion

In other words, `?Sized` fully opts out of all default bounds, and then one has to explicitly opt back in.

```

Author: Fixed

### Adding `?ValueSized`
[adding-valuesized]: #adding-valuesized

Another alternative is to make `ValueSized` a default bound in addition to `Sized` and establish

> that relaxing a supertrait bound also implies relaxing subtrait bounds:

Reviewer: ```suggestion
that relaxing a supertrait bound also implies relaxing subtrait bounds, but relaxing a subtrait bound does
*not* relax the supertraits:
```

Author: Fixed

Prior to the edition migration, the default bound is `ValueSized + const ValueSized + Sized`,

> which could be changed using:

Reviewer: I am surprised by the lack of `const Sized` here. Currently code can call `size_of` in const context on all types, so `const Sized` should be part of the default bound, no?

Author: It should be, revisiting these tables, I felt they were a bit unclear so I've updated them to hopefully be a bit more understandable.

Reviewer: Yes the new tables are much better, thanks :)

| With positive bounds  | Adding `?ValueSized`             |
| --------------------- | -------------------------------- |
| `const Sized`         | `const Sized`                    |
| `Sized`               | `Sized`                          |
| `const ValueSized`    | `?Sized`                         |

> | `ValueSized`          | `?const ValueSized`              |

Reviewer 3: ```suggestion
| `ValueSized`          | `?(const ValueSized)`            |
```

Author: Fixed

| `Pointee`             | `?ValueSized`                    |

After the edition migration, the default bound is
`ValueSized + const ValueSized + Sized + const Sized`, which could be changed using:

| With positive bounds | Adding `?ValueSized`             |
| -------------------- | -------------------------------- |
| `const Sized`        | `const Sized`                    |

> | `Sized`              | `?const Sized`                   |
> | `const ValueSized`   | `?Sized`                         |
> | `ValueSized`         | `?const ValueSized`              |

Reviewer 3: ```suggestion
| `Sized`              | `?(const Sized)`                 |
| `const ValueSized`   | `?Sized`                         |
| `ValueSized`         | `?(const ValueSized)`            |
```

Author: Fixed

      around the issues with constness, but this would have been unsound,
      and ultimately the inability to relax `Clone`'s supertrait made it
      infeasible anyway.

## Why change `size_of` and `size_of_val`?
[why-change-size_of-and-size_of_val]: #why-change-size_of-and-size_of_val

It is theoretically possible for `size_of` and `size_of_val` to accept
runtime-sized types in a const context and use the runtime environment of the host
when computing the size of the types. This has some advantages, if implementable
within the compiler's interpreter, it could enable accelerated execution of
const code. However, there are multiple downsides to allowing this:

Runtime-sized types are often platform specific and could require optional
target features, which would necessitate use of `cfg`s and `target_feature`
with const functions, adding a lot of complexity to const code.

More importantly, the size of a runtime-sized type could differ between
the host and the target and if the size from a const context were to be used
at runtime with runtime-sized types, then that could result in incorrect
code. It is unintuitive that `const { size_of::<svint8_t>() }` would not

> be equal to `size_of::<svint8_t>()`.

Reviewer: It's not just unintuitive, it's unclear how this is even supposed to work. If a `const` has type `(svint8_t, u8)`, and it stores the second field at the offset that it things makes sense, then runtime code will just completely misinterpret this data! So, this alternative has to involve some check ensuring that there is no `svint8_t` in the final value of the constant. Given that Rust has various ways for the programmer to hide types from the compiler, I don't see how this should be possible in a sound way.

Author: Strengthened the wording here to make it clear this isn't possible

          ergonomic rather than avoided.
    - [kennytm][author_kennytm] left a useful comment summarising [which
      standard library bounds would benefit from relaxation to a `DynSized` 
      bound](https://github.com/rust-lang/rust/pull/46108#issuecomment-353672604).
    - Ultimately this was closed [after a language team meeting](https://github.com/rust-lang/rust/pull/46108#issuecomment-360903211)
      deciding that `?DynSized` was ultimately too complex and couldn't be
      justified by support for a relatively niche feature like extern types.
- [rfcs#2255: More implicit bounds (?Sized, ?DynSized, ?Move)][issue_more_implicit_bounds], [kennytm][author_kennytm], Dec 2017
    - Issue created following [rust#46108][pr_dynsized_rebase] to discuss the
      complexities surrounding adding new traits which would benefit from relaxed
      bounds (`?Trait` syntax).
    - There have been various attempts to introduce new auto traits with
      implicit bounds, such as `DynSized`, `Move`, `Leak`, etc. Often rejected due to
      the ergonomic cost of relaxed bounds.
        - `?Trait` being a negative feature can be confusing to users.

>         - Depending on the specific trait being added as a default bound, downstream crates
>           need to re-evaluate every API to determine if adding `?Trait` makes sense, for
>           each `?Trait` added.

Reviewer: ```suggestion
        - Depending on the specific trait being added as a default bound, downstream crates
          need to re-evaluate every API to determine if adding `?Trait` makes sense, for
          each `?Trait` added. (However, the same is true for this RFC, where libraries
          need to determine whether they want to keep or remove the `ValueSized` bounds.)
```

Author: Fixed

>         - `?Trait` isn't actually fully backwards compatible due to interactions with
>           associated types.

Reviewer: It is still not clear to me what this refers to -- I think `?Trait` is exactly as backwards compatible as this proposal. Changing the meaning of existing associated type bounds is not backwards-compatible, but that's a different question. All associated types in old-edition code are `Move` and `DynSized` and `Leak` and whatnot. Changing *that* would be backwards-incompatible. Letting new editions write traits with types that are not `Move` is completely fine, and we can use he `?Move` syntax for that.

So I think this point here should be removed, unless someone can substantiate what the actual claim here is.
```suggestion
```

Reviewer: Alternatively, we could directly discuss this point:
```suggestion
        - existing associated types have to keep their bounds; adding `?Trait` automatically isn't
          backwards compatible. (However, the same happens with this proposal: relaxing associated
          type bounds below their current meaning is not backwards compatible; that's why the edition
          migration has to turn `?Sized` into `const ValueSized`.)
```

Author: Fixed

      `DynSized`) has an explicit bound.
        - This enables deprecation of `?Sized` like this RFC proposes.
    - A `Thin` type to allow thin pointers to DSTs is also proposed in
      this pre-eRFC - it is a different `Thin` from the currently unstable
      `core::ptr::Thin` and it's out-of-scope for this RFC to include a similar type
      and accepted [rfcs#2580][rfc_pointer_metadata_vtable] overlaps.
    - This pre-eRFC may be the origin of the idea for a family of `Sized`
      traits, later cited in [Sized, DynSized, and Unsized][blog_dynsized_unsized].
    - [rfcs#2510][rfc_pointer_metadata_vtable] was later submitted which was a
      subset of this proposal (but none of the `DynSized` parts).
    - This eRFC ultimately fizzled out and didn't seem to result in a proper RFC
      being submitted.
- [rfcs#2310: DynSized without ?DynSized][rfc_dynsized_without_dynsized], [kennytm][author_kennytm], Jan 2018
    - This RFC proposed an alternative version of `DynSized` from
      [rfcs#1993][rfc_opaque_data_structs]/[rust#44469][pr_dynsized] but without
      being an implicit bound and being able to be a relaxed bound (i.e. no
      `?DynSized`).

>         - Adding new implicit bounds which can be relaxed has backwards
>           compatibility hazards, see [rfcs#2255][issue_more_implicit_bounds].

Reviewer: Again I don't understand what those hazards are supposed to be. Adding bounds that *can* be relaxed is entirely fine. Adding bounds that are automatically relaxed for associated types is a problem, but that's a different proposal.

Anyway this point seems redundant here since #2255 is already discussed separately.
```suggestion
```

Author: Fixed

## Alignment
[alignment]: #alignment

There has been community interest in an [`Aligned` trait][rfc_aligned] and there
are examples of `Aligned` traits being added in the ecosystem:

- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
  whether a type has an alignment or not.

An `Aligned` trait hierarchy could be introduced alongside this proposal. It wouldn't
viable to introduce `Aligned` within this hierarchy, as `dyn Trait` which is `ValueSized`
would not be aligned, but some extern types could be `Aligned`, so there isn't an obvious
place that an `Aligned` trait could be included in this hierarchy.

## Custom DSTs
[custom-dsts]: #custom-dsts

Given the community interest in supporting custom DSTs in future (see
[prior art][prior-art]), this RFC was written considering future-compatibility with
custom DSTs in mind.

There are various future changes to these traits which could be used to support
custom DSTs on top of this RFC. None of these have been considered thoroughly, and are
written here only to illustrate.

> - Allow `Pointee` to be implemented manually on user types, which would replace 

Reviewer 4: ```suggestion
- Allow `ptr::Pointee` to be implemented manually on user types, which would replace 
```

Presumably this refers to the existing `Pointee` and not the new one from this RFC, with the idea being that one can define a custom metadata for their custom DST?

Author: It does, clarified this.

                            │ {runtime env, *} │
                            └──────────────────┘
```

Or, in Rust syntax:

```rust
#![feature(const_trait_impl)]

#[const_trait] trait Sized: ~const MetaSized {}

#[const_trait] trait MetaSized: Pointee {}

trait Pointee {}
```

## Implementing `Sized`
[implementing-sized]: #implementing-sized

Implementations of the proposed traits are automatically generated by
the compiler and cannot be implemented manually:

- `Pointee`
    - Types that which can be used from behind a pointer (they may or may
      not have a size).
    - `Pointee` will be implemented for:
        - `MetaSized` and `const MetaSized` types
        - `extern type`s from [rfcs#1861][rfc_extern_types]
        - compound types where every element is `Pointee`
    - In practice, every type will implement `Pointee`.
- `MetaSized`
    - Types whose size is computable given pointer metadata, and knowledge of
      the type, target platform and runtime environment.
      - `const MetaSized` does not require knowledge of the runtime
        environment
    - `MetaSized` is a subtrait of `Pointee`
    - `MetaSized` will be implemented for:
        - `Sized` types

>         - slices `[T]` where every element is `MetaSized`

Reviewer 6: ```suggestion
        - slices `[T]` where every element is `Sized`
```

Looks like a typo...

Author: Fixed in 7a11fe901b0d977e26ee7749b230f05c01ba53bd

which allows operating on multiple values in a single instruction. Processors
have SIMD registers of a known, fixed length and a variety of intrinsics
which operate on these registers. For example, x86-64 introduced 128-bit SIMD
registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.

As an alternative to releasing SIMD extensions with greater bit widths, Arm and
RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
Extension/RVV respectively) where the bit width of vector registers depends on the
CPU implementation, and the instructions which operate these registers are bit
width-agnostic.

As a consequence, these types are not `Sized` in the Rust sense, as the size of
a scalable vector cannot be known during compilation, but is a runtime constant.
For example, the size of these types could be determined by inspecting the value
in a register - this is not available at compilation time and the value may
differ between any given CPU implementation. Both SVE and RVV have mechanisms to
change the system's vector length (up to the maximum supported by the CPU

> implementations) but this is not supported by the proposed ABI for these types.

Reviewer 18: What does `size_of::<ScalableVectorType>()` return if SVE/RVV is not available on target cpu? Would it return `0` or just panic?

Author: Ideally it wouldn't be possible to do this. Initially that could be by having the types behind `cfg`s, but eventually it'll probably require some other RFCs that allows `target_feature` to be used on structs (but not like in the open `struct_target_features` RFC) so that the type can only be used when the target features are available, or something of that sort.

special cases in the type system. See
[rfcs#3268: Scalable Vectors][rfc_scalable_vectors].

## Unsized types and extern types
[unsized-types-and-extern-types]: #unsized-types-and-extern-types

[Extern types][rfc_extern_types] has long been blocked on these types being
neither `Sized` nor `?Sized` ([relevant issue][issue_extern_types_align_size]).
Extern types are listed as a "nice to have" feature in [Rust for Linux's requests
of the Rust project][rfl_want].

RFC #1861 defined that `std::mem::size_of_val` and `std::mem::align_of_val`
should not be defined for extern types but not how this should be achieved, and
suggested an initial implementation could panic when called with extern types,
but this is always wrong, and not desirable behavior in keeping with Rust's
values. `size_of_val` and `align_of_val` both panic for extern types in
the current implementation. Ideally `size_of_val` and `align_of_val` would error
if called with an extern type, but this cannot be expressed in the bounds of
`size_of_val` and `align_of_val` and this remains a blocker for extern types.

Furthermore, unsized types can only be the final member of structs as their

> alignment is unknown and this is necessary to calculate the offsets of later fields.

Reviewer 19: I think you mean size here, not alignment

Author: Fixed

| `Sized`            | Not possible                         | Not possible                         |
| `const ValueSized` | `T: const ValueSized`                | `T: ?Sized`                          |
| `ValueSized`       | `T: ValueSized`                      | Not possible                         |
| `Pointee`          | `T: Pointee`                         | Not possible                         |

After the edition migration, with positive bounds and keeping `?Sized`, the default bound is
`const Sized`, which could be changed using the following syntax:

| Canonically        | Syntax with positive bounds | Syntax keeping `?Sized`                |
| ------------------ | --------------------------- | -------------------------------------- |
| `const Sized`      | `T: const Sized` or `T`     | `T: const Sized` or `T`                |
| `Sized`            | `T: Sized`                  | `T: ?(const Sized) + Sized`            |
| `const ValueSized` | `T: const ValueSized`       | `T: ?(const Sized) + const ValueSized` |
| `ValueSized`       | `T: ValueSized`             | `T: ?(const Sized) + ValueSized`       |

> | `Pointee`          | `T: Pointee`                | `T: ?(const Sized)`                    |

Author: That's true but I've written all of the guide-level and reference-level explanation sections assuming that the future possibilities are not included.

      proposed a bunch of other things necessary for Custom DSTs.
- [rfcs#2310: DynSized without ?DynSized][rfc_dynsized_without_dynsized] was
  proposed at a similar time and was similarly focused only on making `Sized` more
  flexible, but had a bunch of machinery for avoiding backwards incompatibility
  that this RFC believes is unnecessary. Like this proposal, it avoided making
  `DynSized` a default bound and avoided having a relaxed form of it.
    - However, this proposal didn't suggest removing default `Sized` bounds in
      the presence of other size trait bounds.
- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
  major inspiration for this proposal. It has everything this proposal has except
  for the const size traits and all the additional context an RFC needs.

Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
of the type system which can result in new implicit bounds or implicit supertraits
being infeasible for implementation reasons, but [it is believed that this is no
longer relevant][zulip_issue_regions_too_simplistic].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> 

Reviewer 3: It'd make sense to add [this](https://github.com/rust-lang/rfcs/pull/3729#issuecomment-2511408845) as either an unresolved question or future possibility (if it's not already there -- I skimmed but could have missed it):

> ...You're right that `T: MyTrait` would not imply removal of the default bound, it would be `T: const Sized + MyTrait` unless you wrote `T: MyTrait + const ValueSized` (or `T: MyTrait + ?Sized` if we go that way). In [Niko's blog post](https://smallcultfollowing.com/babysteps/blog/2024/04/23/dynsized-unsized/), he suggested it could be that `T: MyTrait` implies removal of the default bound. I mention that what I have proposed [is compatible with that](https://github.com/davidtwco/rfcs/blob/sized-hierarchy/text/3729-sized-hierarchy.md#forward-compatibility-with-supertraits-implying-default-bound-removal) but don't propose it at the moment.

Author: I do mention this in the future possibilities here:
https://github.com/rust-lang/rfcs/blob/462d016a8f8bc00bf0ce6c10e8f52bcf0903cf87/text/3729-sized-hierarchy.md?plain=1#L1684-L1689

elements multiplied by the size of an element - pointer metadata is required to know
how many elements there are, and then information from the runtime environment to
know the size of a `svint8_t` element.

`Pointee` is implemented by any type that can be used behind a pointer, which is
to say, every type (put otherwise, these types may or may not be sized at all).
For example, `Pointee` is therefore implemented on a `u32` which is trivially
sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
no known size.

All type parameters have an implicit bound of `const Sized` which will be
automatically removed if a `~const Sized`, `Sized`, `const MetaSized`,
`~const MetaSized`, `MetaSized` or `Pointee` bound is present instead.

Prior to the introduction of `MetaSized` and `Pointee`, `Sized`'s implicit bound
(now a `const Sized` implicit bound) could be removed using the `?Sized` syntax,
which is now equivalent to a `MetaSized` bound in non-`const fn`s and
`~const MetaSized` in `const fn`s and will be deprecated in the next edition.

> Traits now have an implicit default bound on `Self` of `const MetaSized`.

Author: I've added an edition migration related to this - adding `const MetaSized` to existing traits but not making it an implicit supertrait in the next edition so that traits going forward are implementable on extern types.

its bound should be changed to `T: ~const MetaSized` and this would not result in
any breakage due to the previously described edition migration.

```rust
pub const fn size_of_val<T>(val: &T) -> usize
where
    T: ~const MetaSized,
{
    /* .. */
}
```

While `MetaSized` is equivalent to the current `?Sized` bound it replaces, it
excludes extern types (which `?Sized` by definition cannot), which prevents
`size_of_val` from being called with extern types from [rfcs#1861][rfc_extern_types].
Due to the changes described in [`Sized` bounds][sized-bounds] (migrating
`T: ?Sized` to `T: const MetaSized`), changing the bound of `size_of_val` will
not break any existing callers.

These same changes apply to `align_of` and `align_of_val`.

## Implementing `Copy` for runtime-sized types

Runtime-sized types are value types which should be able to be used as
locals, be copied, etc. To be able to implement [`Copy`][api_copy], these
types would need to implement [`Clone`][api_clone], and to implement
`Clone`, they need to implement `Sized`.

This property trivially falls out of the previous proposals in this RFC:
runtime-sized types are `Sized` (but not `const Sized`) and therefore can

> implement `Clone` and `Copy`.

Reviewer 14: I just chatted at Rust Nation about this. My understanding of the answer to this is: in some sense yes, because locals/arguments/return types are no longer const Sized, but it doesn't depend on the full power of unsized locals (as they're Sized) and it's mostly a case of punting the problem to the backend and LLVM already handles this just fine.

> implement `Clone` and `Copy`.

Author: To clarify further - these types don't need `unsized_locals` machinery - not the skipped `Sized` constraints in the front of the compiler, nor the MIR/codegen changes in the back of the compiler - they just emit different LLVM types and LLVM handles the rest.

> implement `Clone` and `Copy`.

Author: Added some wording clarifying that non-const `Sized` types shouldn't require any special machinery like `unsized_locals`/`unsized_fn_params` to be `Sized`.

| `MetaSized`       | `T: MetaSized`                       | Not possible                         |
| `Pointee`         | `T: Pointee`                         | Not possible                         |

After the edition migration, with positive bounds and keeping `?Sized`, the default bound is
`const Sized`, which could be changed using the following syntax:

| Canonically        | Syntax with positive bounds | Syntax keeping `?Sized`               |
| ------------------ | --------------------------- | ------------------------------------- |
| `const Sized`      | `T: const Sized` or `T`     | `T: const Sized` or `T`               |
| `Sized`            | `T: Sized`                  | `T: ?(const Sized) + Sized`           |
| `const MetaSized`  | `T: const MetaSized`        | `T: ?(const Sized) + const MetaSized` |
| `MetaSized`        | `T: MetaSized`              | `T: ?(const Sized) + MetaSized`       |
| `Pointee`          | `T: Pointee`                | `T: ?(const Sized)`                   |

In other words, `?(const Sized)` fully opts out of all default bounds and then one has to

> explicitly opt back in.

Reviewer 14: I think this was straw-manned slightly in the design meeting, I disagree with the phrasing:

> 'I don't want to think of the "next most precise thing"'

and

> "one level below" in the hierarchy

Neither proposal makes you think like that (ignoring the `?MetaSized` below), the `?Sized` variant requires you to write the positive bound always (and I'd argue that the `T: Pointee` case should be written `T: ?(const Sized) + Pointee`) so the only negative thinking you have to do is "do I need to disable the default `const Sized` bound?".

However, I agree this discussion can be punted until stabilisation given it has essentially no impact on the semantics of what's going on.

P.S The idea of negative trait aliases (`trait Foo = ?Sized;`) terrifies me, it feels like it'd be incredibly confusing at the use-site.

Reviewer 20: Thanks for the pushback @Skepfyr. It's true that we didn't explicitly discuss the idea of writing `?Sized` to "reset" and then a positive bound in combination with it. It's also true that this proposal has many of the advantages of positive bounds — but it's *also* true that it's very verbose and (in my judgment) not the right way forward for that reason. There is definitely a judgment call to be made here, my judgment is that having a small set of "well known" traits that, instead of declared your intent to invoke methods wind up declaring your intent to use it in a more limited way will "wear better" in the long run.

Reviewer 14: I guess my worry is that "small set of "well known" traits" won't be true, with trait aliases and potentially supertrait bounds relaxing the default bound. I agree that `T: ?(const Sized) + Pointee` is horrifically verbose, but if it's just syntax then maybe we could just fix that?

There's nothing stopping us from doing something like one of these:
```rust
fn foo<T: ? + Pointee>() {} // Just drop the (const Sized)
fn foo<T: {} + Pointee>() {} // {} to imply the empty set
fn foo<T: unsized + Pointee>() {} // unsized is a keyword! (probably my favourite)
```

2. Introduce a behaviour where the associated types of subtraits take priority over
  the associated types of supertraits. `Pointee` will effectively become a supertrait
  of all traits so its associated types would never take predecence.
  3. Introduce forward-compatibility lints in current edition, the new traits were
  introduced in the next edition and the edition migration previously described
  in the next next edition.

## Why use const traits?
[why-use-const-traits]: #why-use-const-traits

Previous iterations of this RFC had both linear[^9] and non-linear[^10] trait hierarchies
which included a `RuntimeSized` trait and did not use const traits.

However, both of these were found to be backwards-incompatible due to being unable
to relax the supertrait of `Clone`. In the example below, `Foo::uses_supertrait_bound`'s
call to `requires_sized` relies on `Clone` having a `Sized` supertrait and would no
longer compile if `Clone`'s supertrait was relaxed:

```rust
trait Foo: Clone {
    fn uses_supertrait_bound() { requires_sized::<Self>() }
}

fn requires_sized<T: Sized>() { /* ... */ }
```

Once this interaction with constness was realised, this RFC's dependency on const

> traits was introduced. Due to const traits (at the time of writing this RFC) being

Author: I'm not sure that I agree the const stuff is making the hierarchy complicated, just the nature of these types. Const traits are a natural fit for this and if they were already an accepted and stable part of the language then I don't think there would be any concerns about using them for this proposal.

At a first glance, changing `Clone` like you've suggested seems plausible but I don't think I'll add it as an alternative because of the additional advantages that using const traits give us that changing `Clone` wouldn't, I think those are worth it - these types can't be used in const contexts and we need to be able to prevent that from happening.

> traits was introduced. Due to const traits (at the time of writing this RFC) being

Reviewer 14: I think that's reasonable, I am worried that all of these const traits add too many variations here but you're right that they are correctly representing the types. I was mostly just trying to keep alive the notion that we don't have to perfectly represent all possible variations and if that makes the proposal significantly simpler then it might be a good idea.

          on a new trait after `Pointee` is backwards compatible (except for on a
          trait's associated type). Any existing function will have stricter
          bounds than the new trait and so could be relaxed to the new trait
          without breaking existing code.

## `Sized` bounds
[sized-bounds]: #sized-bounds

`?Sized` would be made syntactic sugar for a `const MetaSized` bound. A
`const MetaSized` bound is equivalent to a `?Sized` bound as all values in Rust
today whose types do not implement `Sized` are valid arguments to
[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
computed given pointer metadata and knowledge of the type and target platform, and
therefore will implement `const MetaSized`. As there are currently no
extern types or other types which would not implement `const MetaSized`,
every type in Rust today which would satisfy a `?Sized` bound would satisfy
a `const MetaSized` bound.

**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
a `const MetaSized` bound. As the `?Trait` syntax is currently accepted for any trait

> but ignored for every trait except `Sized`, `?MetaSized` and `?Pointee` bounds would

Reviewer 3: This is being altered by:

- https://github.com/rust-lang/rust/pull/135841

Author: Thanks - I'll update the wording here.

`Pointee` types cannot be used in non-`#[repr(transparent)]` compound types
as the alignment of these types would need to be known in order to calculate field
offsets. `const Sized` types can be used in compound types with no restrictions.
`Sized`, `const MetaSized` and `MetaSized` types can be used in compound types, but
only as the last element.

## Compiler performance implications
[compiler-performance-implications]: #compiler-performance-implications

There is a potential performance impact within the trait system to adding
supertraits to `Sized`, as implementation of these supertraits will need to be
proven whenever a `Sized` obligation is being proven (and this happens very
frequently, being a default bound). It may be necessary to implement an
optimisation whereby `Sized`'s supertraits are assumed to be implemented and
checking them is skipped - this should be sound as all of these traits are
implemented by the compiler and therefore this property can be guaranteed.

## Forward compatibility with supertraits implying default bound removal
[forward-compatibility-with-supertraits-default-bound]: #forward-compatibility-with-supertraits-implying-default-bound-removal

> Traits which are a supertrait of any of the proposed traits will not
> automatically imply the proposed trait in any bounds where the trait is
> used, e.g.
> 
> ```rust
> trait NewTrait: MetaSized {}

Reviewer 3: ```suggestion
Traits which are a subtrait of any of the proposed traits will not
automatically imply the proposed trait in any bounds where the trait is
used, e.g.

```rust
trait NewTrait: MetaSized {}
```

struct NewRc<T: NewTrait> {} // equiv to `T: NewTrait + Sized` as today
```

> If the user wanted `T: MetaSized` then it would need to be written explicitly.

Reviewer 3: Is this true in just the subtractive direction or in the additive direction as well?  I.e., this works today:

```rust
trait Tr: Sized {}
fn f<T: ?Sized + Tr>() {
    _ = const { size_of::<T>() }
}
```

Author: It would only apply in the subtractive direction (at least w/out the changes you're proposing in [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/410673-t-lang.2Fmeetings/topic/Design.20meeting.202025-02-05/near/499119657) and [referencing here](https://github.com/rust-lang/rfcs/pull/3729#discussion_r1951747465)). I can clarify this.

Author: Added to the example here which should clarify

This is forward compatible with trait bounds which have sizedness supertraits

> implying the removal of the default `const Sized` bound.

Reviewer 3: Is it?  I mean, we accept this today:

```rust
trait Tr /* : ?Sized */ {}
trait Id {
    type Ty: Tr;
}
fn f<T: Id>() {
    _ = const { size_of::<T::Ty>() }
}
```

If we were to use the supertrait's `?Sized` bound to imply the removal of the default `Sized` bound on the associated type, then this would break.  It would seem the same could be said of the new things added here.

Author: It could still be done over an edition, adding `+ Sized` to `T: Id` when we make this change so existing code doesn't break. It's certainly nicer to do it all at once and avoid that, I agree, but I do think it's technically forward compatible. I'm not opposed to making a change like this one as a pre-requisite to this RFC's proposed changes being stabilised, I just think it should be a separate proposal to this one.

Reviewer 3: As a matter of terminology, we generally don't consider a change that requires an edition to be a backward compatible change.  We just say, e.g., that it is a "change we could make over an edition".  Probably the language should be updated here to reflect this.

Specifically, I think the outcome we arrived at is still that we only get one shot, when we stabilize these new bounds, to do this in a way that would work consistently across all editions.  While we could later make things work differently in a new edition, and do a migration dance, that's something we'd generally avoid doing if we already knew we wanted a different behavior.

In terms of the dependencies here, I'm increasingly not a fan of us accepting RFCs with big open questions like this.  Doing so ends up causing us to later stall the stabilization or the work toward it, causing pain.  It's better if we can really nail down the important details, especially those big ones that meaningfully affect how the user would experience the feature.  That way there's a clear path from RFC acceptance to stabilization.

So, it's OK if here we don't want to include this in this RFC, but I'd prefer to see us decide this question ahead of accepting this RFC, and if we decide that we'd like a behavior that isn't fully specified in this RFC, that we then accept another RFC for that ahead of accepting this one, so this one can be updated accordingly and reference it.

Author: > As a matter of terminology, we generally don't consider a change that requires an edition to be a backward compatible change. We just say, e.g., that it is a "change we could make over an edition". Probably the language should be updated here to reflect this.

I don't call this backwards compatible in the RFC, only as "forward compatible", to mean that making a change like this now or later is still possible with this RFC - which is true, we can still make a change like this (w/out an edition now or w/ an edition later) if this RFC is accepted.

Reviewer 3: Right, that's still the nit though.  If starting from some state X, the change Y would not be a backward compatible change, then X is not forward compatible with Y.  If we stabilize these new bounds, then we can't backward-compatibly change the meaning, so changing the meaning is not forward compatible with stabilizing these bounds.

  flexible, but had a bunch of machinery for avoiding backwards incompatibility
  that this RFC believes is unnecessary. Like this proposal, it avoided making
  `DynSized` a default bound and avoided having a relaxed form of it.
    - However, this proposal didn't suggest removing default `Sized` bounds in
      the presence of other size trait bounds.
- [rfcs#3396: Extern types v2][rfc_extern_types_v2] identified that `MetaSized` 
  specifically was necessary moreso than `DynSized` or `ValueSized` and serves
  as the inspiration for this RFC's `MetaSized`.
- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
  major inspiration for this proposal. It has everything this proposal has except
  for the const size traits and all the additional context an RFC needs.

Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
of the type system which can result in new implicit bounds or implicit supertraits
being infeasible for implementation reasons, but [it is believed that this is no
longer relevant][zulip_issue_regions_too_simplistic].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> - What names should be used for the traits?

Reviewer 3: As a note for my fellow @rust-lang/lang members, I'd prefer if we did not leave this question (and perhaps some of the other bikeshed questions in this list) unresolved.

This is in the context of our ongoing discussions about streamlining our feature development processes.

It's always tempting for us to defer these sort of things at RFC time to let work go ahead, but this turns into a false economy, because we then end up getting overloaded and blocking it later, at a worse time.  I'd propose that we just commit to the bikeshed now.

We can always later change our minds, of course.  But we should settle on a presumptive set of choices as part of accepting this RFC so that it has an unencumbered path to stabilization.

Reviewer 7: In light of the strong preference towards positive bounds, I'd like to leave one last comment that I believe it would be very helpful to have a visual indicator for bounds that *relax* requirements rather than adding them.

Moving from `<T>` to `<T: Pointee>` (or `where T: {many bounds}` to `T: {many bounds} + Pointee`) appears to be a breaking change. IMO this will cause significant confusion for library owners, especially as there are several different traits that have this new behavior. This issue could be resolved by introducing a keyword like `<T: keyword Pointee>`, e.g. `only Pointee`, `just Pointee`.

Reviewer 6: > This issue could be resolved by introducing a keyword like `<T: keyword Pointee>`, e.g. `only Pointee`, `just Pointee`.

an existing keyword that might work: `<T: become Pointee>`
`as` or `=` or maybe `@` could also work

There has been community interest in an [`Aligned` trait][rfc_aligned] and there
are examples of `Aligned` traits being added in the ecosystem:

- `rustc` has [its own `Aligned` trait][rustc_aligned] to support pointer tagging.
- [`unsized-vec`][crate_unsized_vec] implements a `Vec` that depends on knowing
  whether a type has an alignment or not.

An `Aligned` trait hierarchy could be introduced alongside this proposal. It wouldn't
viable to introduce `Aligned` within this hierarchy, as `dyn Trait` which is `MetaSized`
would not be aligned, but some extern types could be `Aligned`, so there isn't an obvious
place that an `Aligned` trait could be included in this hierarchy.

## Custom DSTs
[custom-dsts]: #custom-dsts

Given the community interest in supporting custom DSTs in future (see
[prior art][prior-art]), this RFC was written considering future-compatibility with
custom DSTs in mind.

There are various future changes to these traits which could be used to support
custom DSTs on top of this RFC. None of these have been considered thoroughly, and are
written here only to illustrate.

- Allow `std::ptr::Pointee` to be implemented manually on user types, which would

>   replace the compiler's implementation.

Reviewer 21: [normal auto traits (such as Send and Sync) can be explicitly implemented](https://doc.rust-lang.org/beta/unstable-book/language-features/auto-traits.html), the intent of this RFC seems to be that these new traits cannot be (similar to `Sized`).  This should be explicitly noted.

  - Types implementing `Sized` do not require special machinery in the
    compiler, such as `unsized_locals` or `unsized_fn_params`, to be
    considered `Sized`

Introducing new automatically implemented traits is backwards-incompatible,
at least if you try to add it as a bound to an existing function[^2][^3] (and
new auto traits that go unused aren't that useful), but due to being
supertraits of `Sized` and `Sized` being a default bound, these
backwards-incompatibilities are avoided for `SizeOfVal` and `Pointee`.

[^2]: Adding a new automatically implemented trait and adding it as a bound to
      an existing function is backwards-incompatible with generic functions. Even
      though all types could implement the trait, existing generic functions will be
      missing the bound.

      If `Foo` were introduced to the standard library and implemented on every
      type, and it was added as a bound to `size_of` (or any other generic
      parameter)..

      ```rust
      auto trait Foo {}

      fn size_of<T: Sized + Foo>() { /* .. */ } // `Foo` bound is new!
      ```

      ..then user code would break:

      ```rust
      fn do_stuff<T>(value: T) { size_of(value) }

>       // error! the trait bound `T: Foo` is not satisfied

Reviewer 22: this could be avoided using https://lcnr.de/blog/2025/11/28/implicit-auto-traits-assoc-types.html. We could in theory only add a `SizeOfVal` bound on `fn size_of_val` and future compat warn when calling that function with a type containing any generic parameters.

Users then slowly migrate this out. I am not confident that we should do so and whether this is desirable depends on how many users of `?Sized` types actually care about `fn size_of_val`, i.e. how much of a pain is it to propgate these errors outwards.

I also have not fleshed out how this "propagate errors outwards" actually works, e.g. propagating things from impl methods to the trait seems undesirable unless we also add a capability to have traits with optional `SizeOfVal` bounds on some generic parameters This feels related to RTN and being able to say I only support traits where the returned future is `Send`, where here we would say "I also support traits where the function requires `SizeOfVal` (or the other way round 🤷 refinement:tm:)"

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3729/273e41aa87f7e3c364c3b1582f8e6cd66d00507a..1bb22f00880cb6d24192af4041585683b56e9b3a)*

      ```

> [^3]: Trait objects passed by callers would not imply the new trait.

Reviewer 22: not a problem for `SizeOfVal` as all trait objects are `SizeOfVal`

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3729/273e41aa87f7e3c364c3b1582f8e6cd66d00507a..1bb22f00880cb6d24192af4041585683b56e9b3a)*

implementation of this RFC, but bounds in third-party crates need not be.

`Pointee` types will primarily be used for localised FFI and so is not expected
to be so pervasive throughout Rust codebases to the extent that all existing
`?Sized` bounds would need to be immediately reconsidered in light of their
addition, even if in many cases these could be relaxed.

If a user of a `Pointee` type did encounter a bound that needed to be relaxed,
this could be changed in a patch to the relevant crate without breaking
backwards compatibility as-and-when such cases are encountered.

If edition migration were able to attempt migrating each bound to a more relaxed
bound and then use the guaranteed-to-work bound as a last resort then this could
further minimise any changes required by users.

## Other changes to the standard library
[other-changes-to-the-standard-library]: #other-changes-to-the-standard-library

With these new traits and having established changes to existing bounds which
can be made while preserving backwards compatibility, the following changes
could be made to the standard library:

- [`std::boxed::Box`][api_box]

>   - `T: ?Sized` becomes `T: SizeOfVal`

Reviewer 22: so `fn size_of_val` is not the only thing that relies on everything being `SizeOfVal` right now? So does the allocator API and all affected types?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3729/273e41aa87f7e3c364c3b1582f8e6cd66d00507a..1bb22f00880cb6d24192af4041585683b56e9b3a)*

  that this RFC believes is unnecessary. Like this proposal, it avoided making
  `DynSized` a default bound and avoided having a relaxed form of it.
    - However, this proposal didn't suggest removing default `Sized` bounds in
      the presence of other size trait bounds.
- [rfcs#3396: Extern types v2][rfc_extern_types_v2] identified that `SizeOfVal`
  specifically was necessary moreso than `DynSized` or `ValueSized` and serves
  as the inspiration for this RFC's `SizeOfVal`.
- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
  major inspiration for this proposal. It has everything this proposal has
  except for the [future possibilities][future-possibilities] and all the
  additional context an RFC needs.

Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
of the type system which can result in new implicit bounds or implicit supertraits
being infeasible for implementation reasons, but [it is believed that this is no
longer relevant][zulip_issue_regions_too_simplistic].

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- What names should be used for the traits?

>   - One approach would be to name them according to the operation they enable,

Reviewer 6: As mentioned [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/vibeck.20on.20Sized.20Hierarchy.20trait.20names/near/583134570), I think `SizeOfVal` is a bad name since when we add support to `size_of_val` for types that need to read their memory to determine the size (`ValueSized`) rather than just getting the size from the pointer metadata (`MetaSized`), then `SizeOfVal` would change meaning, so I think it's better to stick with a name that directly says the size is based only on pointer metadata, such as `MetaSized`.

I figured I'd mention this here so people who are only following the RFC don't miss that important conversation.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3729/273e41aa87f7e3c364c3b1582f8e6cd66d00507a..1bb22f00880cb6d24192af4041585683b56e9b3a)*

Reviewer 23: I agree completely, except that MetaSized seems a bad way of shortening MetadataSized (metadata is data (about) data, so metasized is sized sized(?) which seems to make no sense). How about PointerSized or RefSized instead?

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Which syntax should be used for opting out of a default bound with const traits and
  a trait hierarchy?
  - This RFC is primarily written proposing the "positive bounds" approach, where
    introducing a positive bound for a supertrait of the default bound will remove
    the default bound.
  - Alternatively, described in [*Adding `?ValueSized`*][adding-valuesized], existing
    relaxed bounds syntax could be used, where a desired bound is written as opting out
    of the next strictest.

> - Should `MetaSized` be introduced?

Reviewer: ```suggestion
- Should `MetaSized` be introduced instead of `ValueSized`?
```
That is the question, right?

Author: We *could* have both, I think replacing `ValueSized` makes sense but I don't think it matters too much to leave that open-ended here.

Reviewer: Right now the RFC sounds like the open question is whether we have just `ValueSized` or both. So I think that should be clarified.

Author: Pushed a commit which should clarify this
